README.md
# muxify

Dynamic Multiplexer

## What is it?

```txt
A Dynamic Multiplexer
     |         |
     |         A gateway to running instances of your server applications
     No need to reloads or restarts. Changes will be applied on the next request
```

### The need

In environments where you might have many Server-Side Rendered (SSR) application instances running on the same machine, and each might listen to a different, and even a random port, popular gateways like `nginx` will not help much, because they cannot get configured during their runtime and they should be restarted after each change in their configuration.

One solution is to have a gateway which distributes requests it receives to application instances that are attached to it, based on a _base URL_ they are interested in. It launches two servers, one is the gateway itself, which routes the requests to application instances. Other one, is the _operation_ server, which application instances can issue requests to it, for attach/detach operations.

Personally, I want to have the following architecture:

![diagram that represent an example scenario where muxify works best](./diagram.png)

*[Credit: <https://app.diagrams.net/>]*

So, if you have the same problem, muxify might help you! 😃

Personally, I use it in development CI/CD environments for multiple instances of server applications.

> It's not fast, and should not be used in production.

### Usage

npx:

  ```sh
  sudo npm install --global muxify
  ```

It will install the `muxify`. Then, you can start it by executing `muxify` command. It will start operation server at `http://127.0.0.3001` and gateway server at `http://127.0.0.3000`.

You can change each server host and port by providing [options](#options) to `muxify`.

### Options

Executing `muxify` with `--help`, will print:

```txt

  Dynamic Multiplexer

  Usage:
    muxify [options]

  Options:
    --operation-port          Preferred operation server port.  [default: 3001]
    --operation-host          Preferred operation server host.  [default: '127.0.0.1']
    --gateway-port            Preferred gateway server port.    [default: 3000]
    --gateway-host            Preferred gateway server host.    [default: '127.0.0.1']
    --help, -h                Prints this help message and exists.
    --version, -v             Prints muxify version (1.0.0-3) and exists.


```

### Operation Server Endpoints

Operation server have two endpoints:

- `PUT /attach`:

  Used to request for attaching a server to a _base URL_.

  Example:

  ```sh
  curl \
      --request PUT \
      --url http://localhost:3001/add-me \
      --header 'Content-Type: application/json' \
      --data '{
          "host": "127.0.0.1",
          "port": 8090,
          "listeningUrl": "/send-me"
      }'
  ```

- `PUT /detach`:

  Used to request for detaching a server from a _base URL_.

  Example:

  ```sh
  curl \
      --request PUT \
      --url http://localhost:3001/add-me \
      --header 'Content-Type: application/json' \
      --data '{
          "listeningUrl": "/im-listening"
      }'
  ```


CHANGELOG.md
# Changelog

## Unreleased

- Remove unnecessary logging

## 1.0.0-3

- Fix bug in pre-release workflow trigger

  There was a bug in pre-release workflow trigger regular expression pattern which triggers the workflow only for tags starting with `v1` or `v2`, not `v3` or `v4`.

- Fix bug in release workflow trigger

  There was a bug in release workflow trigger regular expression pattern which causes the workflow to be run even for pre-release tags, e.g., `v1.0.0-0`, or `v4.6.2-19`. This release fixes this problem and this workflow _should_ only be triggered for stable release tags.

## 1.0.0-2

- Remove publish GitHub Package Registry workflows

  As I realized, YOU CANNOT PUBLISH A PACKAGE BOTH TO GPR AND NPM. Because, GPR is interested in `package.json` `name` field. and tries to match that with the repository, whether owned by you, or a organization (based on your access token). So, the only way you can publish a package both to GPR and NPM is to use a name your package which is the same for NPM and GPR. For example, you should set `"name": "@scope/package-name"` in your `package.json`, where you both have access to _`scope`_ organization on NPM and GitHub (or own the user names, of course).

- Updated outdated dependencies

  - Bump `Rollup` to `2.39.0`.

## 1.0.0-0

- Rename all `dev-mux` occurrences with `muxify`
- Add prepublish workflows for `npm` and `GPR`
- Add prerelease workflow
- Read CLI help message from `package.json` and set it at compile-time
- Update `package.json` `description` field
- Set README.md heading description from `package.json` `description` field

## 0.4.2

- CI/CD:

  Separate publish workflows for more transparency and clearer failures

## 0.4.1

- Fix bug missing binary bundled app to published packages

## 0.4.0

- Rewrite all modules to ES modules `import`/`export` syntax
- Remove unnecessary index.js module. It is replaced by `main.js` module
- Configure rollup to bundle the application
- Git ignore generated binary `bin/muxify`.
  - As it will be generated by CI/CD publish pipelines, there is no need for it to be tracked by git.
- Remove `package.json` `main` field and set `files` array to `bin/muxify`

## 0.3.1

- Auto CHANGELOG.md `Unreleased` header update

  - Add CHANGELOG.md update functionality to post-version script

- Update package.json description

- Update README.md

  - Add documentation on usage and the idea behind muxify

- Add README.md auto help message update

  - Create `readme-template.md` which is a pre-processed version of README.md

## 0.3.0

- Breaking Changes:

  - `PUT /add-me` route URL changed to `PUT /attach`. It accepts the same payload as previous, only route name changed due to name consistency with newly added `/detach` route URL.

- Features:

  - Add detach server URL which can be used by servers providing the url they were listening to, to detach them from proxy table. An example of such a request for a server that was listening to requests wit base URL `/im-listening` to operation server listening at `http://localhost:3001`, will be like following:

    ```sh
    curl \
        --request PUT \
        --url http://localhost:3001/add-me \
        --header 'Content-Type: application/json' \
        --data '{
            "listeningUrl": "/im-listening"
        }'
    ```

- Bug Fixes:

  - Fix bug adding proxy rule on all events emitted by event emitter in gateway server.

## 0.2.7

- CI/CD:

  - Remove `scope` parameter from _publish GPR_ job
  - Add _Upgrade NPM_ step to _publish GPR_ job in order to fix issue in authentication.

    Previous versions of npm had bug reading `_auth` key from `.npmrc` (which is set by [actions/setup-node](https://github.com/actions/setup-node)), as mentioned in [1](https://github.com/actions/setup-node/issues/213#issuecomment-752095615), and [2](https://github.com/npm/cli/issues/2300).

## 0.2.6

- CI/CD:

  - Remove GitHub registry URL auth token configuration to `.npmrc`
  - Add `scope` parameter to _publish GPR_ job

## 0.2.5

- CI/CD:
  - Add GitHub registry URL auth token configuration to `.npmrc`
  - Revert _publish_ workflow _publish GPR_ job node version back to `15`

## 0.2.4

- CI/CD:
  - Change _publish_ workflow _publish GPR_ job node version to `14`

## 0.2.3

- CI/CD:

  - Add on release triggers to _publish_ workflow configuration

  - Change publish GPR token to virtual environment provided GITHUB_TOKEN

## 0.2.2

- CI/CD:

  Set Git default branch to `main` in _publish_ workflow.

## 0.2.1

- CI/CD:

  Add more triggers to _publish_ workflow.

## 0.2.0

- Add ability to register a server:

  Servers can register themselves to gateway proxy table providing the options required to register them. The request payload consists of the URL they want to receive requests on (`listeningUrl`), `host` and `port` of their listener server. Example of a request for a gateway with operation server listening at `http://localhost:3001` for a server listening at `http://127.0.0.1:8090` that wants to receive requests with the base URL `/send-me` will be:

  ```sh
  curl \
      --request PUT \
      --url http://localhost:3001/add-me \
      --header 'Content-Type: application/json' \
      --data '{
          "host": "127.0.0.1",
          "port": 8090,
          "listeningUrl": "/send-me"
      }'
  ```


src/main.js
import { GatewayServer } from './gateway-server';
import { OperationServer } from './operation-server';
import { EventEmitter } from 'events';
import minimist from 'minimist';


const help = `
  process.env.DESCRIPTION

  Usage:
    muxify [options]

  Options:
    --operation-port          Preferred operation server port.  [default: 3001]
    --operation-host          Preferred operation server host.  [default: '127.0.0.1']
    --gateway-port            Preferred gateway server port.    [default: 3000]
    --gateway-host            Preferred gateway server host.    [default: '127.0.0.1']
    --help, -h                Prints this help message and exists.
    --version, -v             Prints muxify version (process.env.VERSION) and exists.
`;

const argv = minimist(
  process.argv.slice(2),
  {
    string: ['operation-host', 'gateway-host'],
    boolean: ['help', 'version'],
    alias: {
      help: ['h'],
      version: ['v']
    },
    default: {
      help: false,
      version: false,
      'operation-port': 3001,
      'operation-host': '127.0.0.1',
      'gateway-port': 3000,
      'gateway-host': '127.0.0.1',
    },
    unknown: (unknownArgument) => {
      console.error(`Unknown argument '${unknownArgument}'`);
      console.error(help);
      process.exit(1);
    },
    stopEarly: true
  },
);

const flags = {
  operationPort: argv['operation-port'],
  operationHost: argv['operation-host'],
  gatewayPort: argv['gateway-port'],
  gatewayHost: argv['gateway-host'],
  help: argv.help,
  version: argv.version,
}

if (flags.help) {
  console.log(help);
  process.exit(0);
}

if (flags.version) {
  console.log('process.env.VERSION');
  process.exit(0);
}

const eventEmitter = new EventEmitter();

const gatewayServer = GatewayServer(eventEmitter);
gatewayServer.start(
  flags.gatewayHost,
  flags.gatewayPort,
  () => {
    console.info(`muxify gateway server is listening at http://${flags.gatewayHost}:${flags.gatewayPort}`);
  }
);

const operationServer = OperationServer(eventEmitter);
operationServer.start(
  flags.operationHost,
  flags.operationPort,
  () => {
    console.info(`muxify operation server is listening at http://${flags.operationHost}:${flags.operationPort}`);
  }
);
